extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/PanacheQuery.java

```diff
@@ -0,0 +1,170 @@
+package io.quarkus.mongodb.panache;
+
+import java.util.List;
+import java.util.stream.Stream;
+
+import io.quarkus.panache.common.Page;
+
+/**
+ * <p>
+ * Interface representing an entity query, which abstracts the use of paging, getting the number of results, and
+ * operating on {@link List} or {@link Stream}.
+ * </p>
+ * <p>
+ * Instances of this interface cannot mutate the query itself or its parameters: only paging information can be
+ * modified, and instances of this interface can be reused to obtain multiple pages of results.
+ * </p>
+ *
+ * @author Loïc Mathieu - loikeseke@gmail.com
+ * @param <Entity> The entity type being queried
+ */
+public interface PanacheQuery<Entity> {
+
+    // Builder
+
+    /**
+     * Sets the list of projection fields.
+     * The query will only retrieve these fields from the database.
+     * Note that the _id field will always be returned.
+     *
+     * @return this query, modified
+     */
+    public <T extends Entity> PanacheQuery<T> project(String firstField, String... fields);
```

Comment:
No, it just record in the PanacheQuery that it needs to be done with projection fields. So it must be used before calling the terminator operations list/stream.

Reply:
https://www.baeldung.com/jpa-hibernate-projections says projections typically alter the type of objects returned by your query. So I assume `project(String oneField)` returns a `PanacheQuery<Object>` and `project(String oneField, String otherField)` returns a `PanacheQuery<Object[]>` or something like it. And that's not even mentioning that you can agregate multiple fields into another type of object by invoking its constructor.

Was it @emmanuelbernard who made you add projections? If yes we should check that what they do and he thinks they should do have some intersection ;)

Reply:
I understand that having null in fields can be disturbing and that 'semantically' a projection should return another type of Object.
Ideally, a user should create a separate entity mapped to the same table and use it to only return the needed fields, or use a separated interface (this is how it is done with spring data mongodb).

The way I did it is simpler and more pragmatic : you list the column you want to project on and only those will be retrieved from the db. 

I really want to avoid forcing the user working with Obect as you suggest ...

Reply:
Again, I don't really have an opinion until I know what @emmanuelbernard wanted :)

Reply:
I'm not sure I like that form of projection (which is really avoiding loading some properties).

BTW unrelated, can you have nested objects and collections of objects in the BSON ish mapping?

Reply:
BSON Document supports nested objects and array of objects (I didn't use a lot arrays of objects). 
Nested objects are the prefered way to describe relations between two objects in the NoSQL world (no relation between collections so store everything in a single collection)